core/model/modx/sources/mods3mediasource.class.php

```diff
@@ -47,181 +41,298 @@ public function initialize() {
                 'lexicon' => 'core:source',
-            ),
-            'secret_key' => array(
+            ],
+            'secret_key' => [
                 'name' => 'secret_key',
                 'desc' => 'prop_s3.secret_key_desc',
                 'type' => 'password',
                 'options' => '',
                 'value' => '',
                 'lexicon' => 'core:source',
-            ),
-            'imageExtensions' => array(
+            ],
+            'imageExtensions' => [
                 'name' => 'imageExtensions',
                 'desc' => 'prop_s3.imageExtensions_desc',
                 'type' => 'textfield',
                 'value' => 'jpg,jpeg,png,gif,svg',
                 'lexicon' => 'core:source',
-            ),
-            'thumbnailType' => array(
+            ],
+            'thumbnailType' => [
                 'name' => 'thumbnailType',
                 'desc' => 'prop_s3.thumbnailType_desc',
                 'type' => 'list',
-                'options' => array(
-                    array('name' => 'PNG','value' => 'png'),
-                    array('name' => 'JPG','value' => 'jpg'),
-                    array('name' => 'GIF','value' => 'gif'),
-                ),
+                'options' => [
+                    ['name' => 'PNG', 'value' => 'png'],
+                    ['name' => 'JPG', 'value' => 'jpg'],
+                    ['name' => 'GIF', 'value' => 'gif'],
+                ],
                 'value' => 'png',
                 'lexicon' => 'core:source',
-            ),
-            'thumbnailQuality' => array(
+            ],
+            'thumbnailQuality' => [
                 'name' => 'thumbnailQuality',
                 'desc' => 'prop_s3.thumbnailQuality_desc',
                 'type' => 'textfield',
                 'options' => '',
                 'value' => 90,
                 'lexicon' => 'core:source',
-            ),
-            'visibility' => array(
+            ],
+            'visibility' => [
                 'name' => 'visibility',
                 'desc' => 'prop_file.visibility_desc',
-                'type' => 'textfield',
+                'type' => 'modx-combo-visibility',
                 'options' => '',
                 'value' => 'public',
                 'lexicon' => 'core:source',
-            ),
-            'skipFiles' => array(
+            ],
+            'skipFiles' => [
                 'name' => 'skipFiles',
                 'desc' => 'prop_s3.skipFiles_desc',
                 'type' => 'textfield',
                 'options' => '',
                 'value' => '.svn,.git,_notes,nbproject,.idea,.DS_Store',
                 'lexicon' => 'core:source',
-            ),
-        );
+            ],
+        ];
     }
 
+
+    /**
+     * @return array
+     */
+    public function getListDirContextMenu()
+    {
+        $menu = parent::getListDirContextMenu();
+        foreach ($menu as $k => $v) {
+            if ($v['handler'] == 'this.renameDirectory') {
```

Comment:
Interesting approach to disallowing the rename on s3 directories - assuming you checked if there were a cleaner way to prevent that option?

Reply:
Another option is to do not use the parent method and generate the whole menu in S3 class. I think, it is better to just remove one item.